homeassistant/components/coinbase/sensor.py

```diff
@@ -16,45 +28,70 @@
 
 ATTRIBUTION = "Data provided by coinbase.com"
 
-DATA_COINBASE = "coinbase_cache"
 
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Coinbase sensor platform."""
+    instance = hass.data[DOMAIN][config_entry.entry_id]
+    hass.async_add_executor_job(instance.update)
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up the Coinbase sensors."""
-    if discovery_info is None:
-        return
-    if "account" in discovery_info:
-        account = discovery_info["account"]
-        sensor = AccountSensor(
-            hass.data[DATA_COINBASE], account["name"], account["balance"]["currency"]
-        )
-    if "exchange_currency" in discovery_info:
-        sensor = ExchangeRateSensor(
-            hass.data[DATA_COINBASE],
-            discovery_info["exchange_currency"],
-            discovery_info["native_currency"],
-        )
+    entities = []
+
+    if CONF_CURRENCIES in config_entry.options:
+        desired_currencies = config_entry.options[CONF_CURRENCIES]
+    else:
+        desired_currencies = [
+            account[API_ACCOUNT_CURRENCY] for account in instance.accounts
+        ]
+
+    exchange_native_currency = instance.exchange_rates.currency
+
+    for currency in desired_currencies:
+        entities.append(AccountSensor(instance, currency))
+
+    if CONF_EXCHANGE_RATES in config_entry.options:
+        for rate in config_entry.options[CONF_EXCHANGE_RATES]:
+            entities.append(
+                ExchangeRateSensor(
+                    instance,
+                    rate,
+                    exchange_native_currency,
+                )
+            )
 
-    add_entities([sensor], True)
+    async_add_entities(entities)
 
 
 class AccountSensor(SensorEntity):
     """Representation of a Coinbase.com sensor."""
 
-    def __init__(self, coinbase_data, name, currency):
+    def __init__(self, coinbase_data, currency):
         """Initialize the sensor."""
         self._coinbase_data = coinbase_data
-        self._name = f"Coinbase {name}"
-        self._state = None
-        self._unit_of_measurement = currency
-        self._native_balance = None
-        self._native_currency = None
+        self._currency = currency
+        for account in coinbase_data.accounts:
+            if account.currency == currency:
```

Comment:
What happens if the currency is no longer available at this point?

Reply:
Currently we fail silently and don't add the sensor. Will explicitly throw a warning

Reply:
Hmz yeah...
Not sure, if it is a temporary error, it should still become a sensor, but marked as unavailable. If this happens because of currency availability removed, maybe the sensor should be cleaned up.

Thinking about it, it might be better to move this logic into the platform setup, so it gets figured out before the sensor is instantiated even.

Reply:
Yep I found in the case where we were init'ing a non-provided sensor we found fail setup. So now the logic is moved up so we don't try and create an unprovided sensor at all